### PR TITLE
feat: migrate image list 

### PR DESCRIPTION
**Pull request by gahyuun**
### TL;DR

migrate ImageList 

###  Why make this change?

To migrate from lit component to react component

### What changed?

- ImageInstallModal
    - Component that install images
    - I deselected the rows when clicking the install button in the modal because I felt the flow of deselecting the rows was more natural, whereas before it would not deselect the rows.
    - The newly added phrases are written in en and ko only
    - However, I didn't proceed with replacing "download" with "install". This was because there were more download phrases than I thought, and I thought it might be more confusing if I replaced a lot of them

- ImageList
    - I used suspense to prevent the entire screen from loading when the image list is first rendered

- ManageAppsModal, ManageImageResourceLimitModal
    - I modified it because it's called from a React component, not a lit component.
    - Implemented the getServicePorts function in ManageAppsModal without passing service port as a prop 

- ResourceNumber
    - I added ranges via the range and max properties ( ex 1~∞ Core )
    - The ResourceNumber component made rendering the table too slow, so I used Memo

### How to test?

- imageList
1. enter the environment page 
2. select the row
3. sort the column

- ManageImageResourceLimitModal
1. click button
3. change resource
4. click Okay button
5. check if the resource has changed

- ManageAppsModal
1. click button
2. add port
3. delete port
4. check if the ports has changed

- ImageInstallModal
1. select uninstalled images
2. click install button
3. check uninstalled images info
4. click install button and check if an notification appears

1. select installed images
2. check warn message

1. select uninstalled images and installed images
2. check exclude message

### CODE REVIEW

File: react/src/components/ImageList.tsx [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        );
+      },
+    },
+  ];
+
+  return (
+    <>
+      <Flex justify="end" style={{ height: '3rem', padding: '.5rem 1rem' }}>
+        <Button
+          size="middle"
```
Comment (ironAiken2) [inline react/src/components/ImageList.tsx]: ant'd button's default size is 'middle'

File: react/src/index.tsx [hunk 2]
```diff
@@ -284,13 +256,11 @@ const root = ReactDOM.createRoot(
   document.getElementById('react-root') as HTMLElement,
 );
 root.render(
-  <React.StrictMode>
```
Comment (ironAiken2) [inline react/src/index.tsx]: Why did you remove this code?

File: react/src/pages/EnvironmentPage.tsx [hunk 3]
```diff
@@ -45,38 +62,40 @@ const EnvironmentPage = () => {
         },
       }}
     >
+      {curTabKey === 'image' ? (
+        <Flex
+          style={{
+            display: 'block',
+            paddingTop: token.paddingContentVerticalSM,
```
Comment (ironAiken2) [inline react/src/pages/EnvironmentPage.tsx]:
```suggestion
```

This code seems to break the consistency of the css styles. It would be better to control it inside the `ImageList`.

File: react/src/components/ImageList.tsx [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+import _ from 'lodash';
+import { Key, useMemo, useState, useTransition } from 'react';
+import { useTranslation } from 'react-i18next';
+import { useLazyLoadQuery } from 'react-relay';
+
+export type EnvironmentImage = NonNullable<
+  ImageListQuery$data['images']
+>[number];
+
+export default function ImageList() {
```
Comment (ironAiken2) [inline react/src/components/ImageList.tsx]: I think it would be better to use the arrow function. And it would be nice to be able to pass the style to props in case we want to change the style of that component elsewhere. 
```suggestion
const ImageList: React.FC<{ style?: React.CSSProperties;}> = ({ style }) => {
```

File: react/src/components/ImageList.tsx [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+            </Flex>
+          </>
+        );
+      },
+    },
+  ];
+
+  return (
+    <>
+      <Flex justify="end" style={{ height: '3rem', padding: '.5rem 1rem' }}>
```
Comment (ironAiken2) [inline react/src/components/ImageList.tsx]:
```suggestion
    <Flex
      direction="column"
      align="stretch"
      style={{
        flex: 1,
        ...style,
      }}
    >
      <Flex justify="end" style={{ padding: token.paddingSM }}>
```

It seems to be changed to maintain stylistic consistency with other components on the same page.

File: react/src/components/ImageList.tsx [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+          onClick={() => {
+            setIsOpenInstallModal(true);
+          }}
+        >
+          {t('environment.Install')}
+        </Button>
+      </Flex>
+      <Table
+        rowKey="id"
+        scroll={{ x: '100vw', y: 'calc(100vh - (9.0625rem + 3rem + 4.5rem))' }}
```
Comment (ironAiken2) [inline react/src/components/ImageList.tsx]:
```suggestion
scroll={{ x: 'max-content' }}
```

It seems to be affecting the styling of the table columns. The following modifications should fix the issue.

**Review by ironAiken2 — CHANGES_REQUESTED** [review]
Currently, we don't have pagination, so It's getting too much data and it's causing lag on page navigation and other interactions. Can you take a look at this as well?

File: react/src/index.tsx (hunk 2 above)
Comment (gahyuun) [reply react/src/index.tsx]: my misake .. i modified it

File: react/src/components/ImageList.tsx (hunk 4 above)
Comment (gahyuun) [reply react/src/components/ImageList.tsx]: I fixed style related issues and added the style prop

Comment (gahyuun) [reply react/src/components/ImageList.tsx]: I used function declaration to write the declaration and default export at once. can you explain why arrow function is better in react component?

File: react/src/components/ImageInstallModal.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  };
+  const { hasInstalledImage, imagesToInstall } = mapImages();
+
+  if (imagesToInstall.length === 0) {
+    message.warning(t('environment.AlreadyInstalledImage'));
+    onRequestClose();
+    return null;
+  }
+
+  const handleClick = () => {
```
Comment (gahyuun) [inline react/src/components/ImageInstallModal.tsx:39]: handleClick references the _installImage function in the lit code 
https://github.com/lablup/backend.ai-webui/blob/main/src/components/backend-ai-environment-list.ts#L618

File: react/src/components/ImageList.tsx [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    [images],
+  );
+  const columns: ColumnsType<EnvironmentImage> = [
+    {
+      title: t('environment.Status'),
+      dataIndex: 'installed',
+      key: 'installed',
+      sorter: sortBasedOnInstalled,
+      render: (text, row) => {
+        if (row?.id && installingImages.includes(row.id))
```
Comment (gahyuun) [inline react/src/components/ImageList.tsx]: I used the installingImages state to show the installing tag

File: react/src/components/ImageList.tsx [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+            }}
+          >
+            {t('environment.Install')}
+          </Button>
+        </Flex>
+        <Table
+          rowKey="id"
+          scroll={{
+            x: 'max-content',
+            y: 'calc(100vh - (9.0625rem + 3rem + 4.5rem))',
```
Comment (gahyuun) [inline react/src/components/ImageList.tsx:327]: The reason I specified scroll on the y-axis is that I thought the colum header and install button should be visible while scrolling.

File: react/src/components/ImageList.tsx [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+  };
+
+  const sortBasedOnInstalled = (a: EnvironmentImage, b: EnvironmentImage) => {
+    return a?.installed && !b?.installed
+      ? -1
+      : !a?.installed && b?.installed
+        ? 1
+        : 0;
+  };
+
```
Comment (gahyuun) [inline react/src/components/ImageList.tsx]: https://github.com/lablup/backend.ai-webui/blob/main/src/components/backend-ai-environment-list.ts#L461
lit code for reference

File: react/src/components/ImageTags.tsx [hunk 11]
```diff
@@ -85,48 +84,49 @@ interface LangTagsProps extends TagProps {
 }
 export const LangTags: React.FC<LangTagsProps> = ({ image, ...props }) => {
   image = image || '';
-  const [, { getImageLang, tagAlias }] = useBackendAIImageMetaData();
+  const [, { getImageLang }] = useBackendAIImageMetaData();
   return (
     <Tag color="green" {...props}>
-      {tagAlias(getImageLang(image))}
+      {getImageLang(image)}
```
Comment (gahyuun) [inline react/src/components/ImageTags.tsx:90]: The reason for removing tagalias is that sometimes it is necessary to show the language without using the LangTags component, so I think it is better to include tagAlias inside the getImageLang function.

File: resources/image_metadata.json
Comment (gahyuun) [inline resources/image_metadata.json:1]: While checking ImageList, I added the parts that didn't have metaData.

File: react/src/hooks/index.tsx [hunk 12]
```diff
@@ -251,7 +260,49 @@ export const useBackendAIImageMetaData = () => {
         const { tags } = getImageMeta(imageName);
         return tags[1];
       },
+      getBaseImages: (tag: string, name: string) => {
```
Comment (gahyuun) [inline react/src/hooks/index.tsx:268]: `getBaseImages`,`getConstraints`,`getLangs` implemented a new function without modifying the `getBaseImage`,`getCustomTag`,`getFilteredRequirementsTag`,`getImageLang` because of side effects ( ex. `SessionKernelTags` uses `getBaseImage`. )
Also, because of the side effect, I didn't modify `MyEnvironmentPage` to any values except the constraints. I think it's better to do this in a new task.

File: react/src/hooks/index.tsx
Comment (gahyuun) [inline react/src/hooks/index.tsx:1]: https://github.com/lablup/backend.ai-webui/blob/main/src/components/backend-ai-environment-list.ts#L345
lit code for reference

**Review by ironAiken2 — CHANGES_REQUESTED** [review]
Please resolve conflict :)

File: react/src/hooks/useBackendAIImageMetaData.test.tsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            r: 'R',
+            'r-base': 'R',
+            py3: 'Python3',
+            ngc: 'Nvidia GPU Cloud',
+            py310: 'Python3',
+          },
+          tagReplace: {},
+        }),
+    }),
+  );
```
Comment (gahyuun) [inline react/src/hooks/useBackendAIImageMetaData.test.tsx:31]: I used jest.fn because suspense can't wait for the test to complete while fetching data

File: react/src/pages/EnvironmentPage.tsx [hunk 14]
```diff
@@ -50,12 +51,15 @@ const EnvironmentPage = () => {
     >
       <Flex
         style={{
-          display: curTabKey === 'image' ? 'block' : 'none',
-          paddingTop: token.paddingContentVerticalSM,
+          display: 'block',
         }}
       >
```
Comment (ironAiken2) [inline react/src/pages/EnvironmentPage.tsx]: This seems to be a remnant of code written to manage the display of the web component, how about removing this line?

File: react/src/components/ImageInstallModal.tsx [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+
+    return { hasInstalledImage, imagesToInstall };
+  };
+  const { hasInstalledImage, imagesToInstall } = mapImages();
+
+  if (imagesToInstall.length === 0) {
+    message.warning(t('environment.AlreadyInstalledImage'));
+    onRequestClose();
+    return null;
+  }
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx]: If the modal's openness were managed inside the modal, it would consume unnecessary resources, and since it is not the role of the component, it seems better to manage it in the parent component.

File: react/src/components/ImageList.tsx [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+          ...style,
+        }}
+      >
+        <Flex justify="end" style={{ padding: token.paddingSM }}>
+          <Button
+            icon={<VerticalAlignBottomOutlined />}
+            style={{ backgroundColor: token.colorPrimary, color: 'white' }}
+            onClick={() => {
+              setIsOpenInstallModal(true);
+            }}
```
Comment (ironAiken2) [inline react/src/components/ImageList.tsx]: [related comment](https://app.graphite.dev/github/pr/lablup/backend.ai-webui/2615/feat-migrate-image-list#comment-PRRC_kwDOCRTcws5mG_cB)

File: react/src/components/ImageInstallModal.tsx [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+  const { upsertNotification } = useSetBAINotification();
+  const painKiller = usePainKiller();
+
+  if (!open) return null;
+
+  if (selectedRowKeys.length === 0) {
+    message.warning(t('environment.NoImagesAreSelected'));
+    onRequestClose();
+    return null;
+  }
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx]: Same as [related comment](https://app.graphite.dev/github/pr/lablup/backend.ai-webui/2615/feat-migrate-image-list#comment-PRRC_kwDOCRTcws5mG_cB)

File: react/src/components/ImageInstallModal.tsx (hunk 15 above)
Comment (gahyuun) [reply react/src/components/ImageInstallModal.tsx:61]: However, if the parent component manages the determination of whether an image is already installed, then mapImages() must be implemented in the parent component as well. 
What do you think about this resource?

File: react/src/components/ImageInstallModal.tsx [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+        .map((item) => item?.id)
+        .filter((id): id is string => id !== null && id !== undefined),
+    );
+    setSelectedRowKeys([]);
+  };
+
+  return (
+    <BAIModal
+      destroyOnClose
+      open={open}
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx]:
```suggestion

```

File: react/src/components/ImageInstallModal.tsx [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+  setSelectedRowKeys: Dispatch<SetStateAction<Key[]>>;
+}
+export default function ImageInstallModal({
+  open,
+  onRequestClose,
+  selectedRowKeys,
+  images,
+  setInstallingImages,
+  setSelectedRowKeys,
+}: ImageInstallModalInterface) {
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx]:
```suggestion
export default function ImageInstallModal({
  onRequestClose,
  selectedRowKeys,
  images,
  setInstallingImages,
  setSelectedRowKeys,
  ...modalProps
}: ImageInstallModalInterface) {
```

File: react/src/components/ImageInstallModal.tsx [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+  return (
+    <BAIModal
+      destroyOnClose
+      open={open}
+      maskClosable={false}
+      onCancel={() => onRequestClose()}
+      title={t('dialog.title.LetsDouble-Check')}
+      okText={t('environment.Install')}
+      onOk={handleClick}
+    >
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx:153]:
```suggestion
      onOk={handleClick}
      {...modalProps}
    >
```

File: react/src/components/ImageInstallModal.tsx [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+import { useTranslation } from 'react-i18next';
+
+interface ImageInstallModalInterface extends BAIModalProps {
+  open: boolean;
+  onRequestClose: () => void;
+  selectedRowKeys: Key[];
+  images: EnvironmentImage[];
+  setInstallingImages: Dispatch<SetStateAction<string[]>>;
+  setSelectedRowKeys: Dispatch<SetStateAction<Key[]>>;
+}
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx]: If you declare a new interface by extending antd's default modal, the fields included in the existing modal interface can be removed from the declaration and also other fields in the main modal interface are also available.[[related comment]](https://app.graphite.dev/github/pr/lablup/backend.ai-webui/2615/feat-migrate-image-list#comment-PRRC_kwDOCRTcws5mHAhD) [[related comment]](https://app.graphite.dev/github/pr/lablup/backend.ai-webui/2615/feat-migrate-image-list#comment-PRRC_kwDOCRTcws5mHAoW) [[related comment]](https://app.graphite.dev/github/pr/lablup/backend.ai-webui/2615/feat-migrate-image-list#comment-PRRC_kwDOCRTcws5mHAtJ)

File: react/src/components/ImageInstallModal.tsx [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+              ),
+          );
+          // @ts-ignore
+          globalThis.lablupNotification.text = painKiller.relieve(error.title);
+          // @ts-ignore
+          globalThis.lablupNotification.detail = error.message;
+          // @ts-ignore
+          globalThis.lablupNotification.show(true, error);
+          indicator.set(100, t('environment.DescProblemOccurred'));
+          indicator.end(1000);
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx]: Is there a reason why you didn't use `upsertNotification` in this part?

File: react/src/components/ImageInstallModal.tsx [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+        .then(() => {
+          indicator.end(1000);
+        })
+        .catch((error: any) => {
+          setInstallingImages(
+            imagesToInstall
+              .map((item) => item?.id)
+              .filter(
+                (id): id is string =>
+                  id !== null && id !== undefined && id !== image?.id,
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx:146]: [ref](https://github.com/lablup/backend.ai-webui/pull/2615/files#diff-8ee05aa0711da52a475b27ba894a607046f46a5c4cf64903702679637cefbef4R141-R148) Isn't this duplicated?

File: react/src/components/ImageInstallModal.tsx [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+          globalThis.lablupNotification.show(true, error);
+          indicator.set(100, t('environment.DescProblemOccurred'));
+          indicator.end(1000);
+        });
+    });
+    setInstallingImages(
+      imagesToInstall
+        .map((item) => item?.id)
+        .filter((id): id is string => id !== null && id !== undefined),
+    );
```
Comment (ironAiken2) [inline react/src/components/ImageInstallModal.tsx]: Isn't it duplicated with [ref](https://github.com/lablup/backend.ai-webui/pull/2615/files#diff-8ee05aa0711da52a475b27ba894a607046f46a5c4cf64903702679637cefbef4R159-R164)? And it seems that the problem will occur if you download another image additionally without refreshing the page.

File: react/src/components/ImageList.tsx [hunk 25]
```diff
@@ -0,0 +1,369 @@
...
+        <Table
+          rowKey="id"
+          scroll={{
+            x: 'max-content',
+            y: 'calc(100vh - (9.0625rem + 3rem + 4.5rem))',
+          }}
+          pagination={false}
+          dataSource={sortedImages}
+          columns={columns as ColumnType<AnyObject>[]}
+          rowSelection={rowSelection}
```
Comment (ironAiken2) [inline react/src/components/ImageList.tsx]:
```suggestion
          rowSelection={{
            type: 'checkbox',
            onChange: (_, selectedRows: DataType[]) => {
              setSelectedImage(selectedRows);
            },
          }}
```

I think we could save the selected image as a state, and pass only that image to installmodal, to reduce unnecessary filtering logic.

File: react/src/components/ImageInstallModal.tsx (hunk 18 above)
Comment (gahyuun) [reply react/src/components/ImageInstallModal.tsx:170]:
```
 <BAIModal
      destroyOnClose
      open
..
```
Is this code suggested?

File: react/src/components/FlexActivityIndicator.tsx
Comment (gahyuun) [inline react/src/components/FlexActivityIndicator.tsx:1]: I modified it because I wanted to resize the spinner

File: react/src/components/ImageList.tsx [hunk 26]
```diff
@@ -0,0 +1,414 @@
...
+        display: 'flex',
+        alignItems: 'center',
+        height: '100%',
+        ...style,
+      }}
+    >
+      {children}
+    </div>
+  );
+};
```
Comment (gahyuun) [inline react/src/components/ImageList.tsx]: In Ant Design's Table component, when using virtual mode and specifying a scroll.y value, the vertical center alignment does not seem to be applied correctly because the cells have a fixed height. 
So, I implemented a component called CellWrapper and did the vertical alignment myself.  

Since I was doing a custom render, the ellipsis property was not applied, so I applied ellipsis directly as well

File: react/src/components/ImageList.tsx [hunk 27]
```diff
@@ -0,0 +1,414 @@
...
+            }}
+          >
+            {t('environment.Install')}
+          </Button>
+        </Flex>
+        <Table<EnvironmentImage>
+          rowKey="id"
+          scroll={{
+            x: 1600,
+            y: window.innerHeight - 145 - 56 - 54,
```
Comment (gahyuun) [inline react/src/components/ImageList.tsx]: 145: ref ) https://github.com/lablup/backend.ai-webui/pull/2615/files#diff-7513e7364d1bbef624ab929ba47fa51f2fcb5c0090f36db4f56e3837bb29bac5R78

56 : The space where the install button exists 

54: column title of table

File: react/src/components/ImageInstallModal.tsx (hunk 22 above)
Comment (gahyuun) [reply react/src/components/ImageInstallModal.tsx]: - The way to display the error in upsertNotification seemed to be to set the backgroundTask status to rejected.
But there is no backgroundTask.
- I also referenced the container Registry List code 
ref) https://github.com/lablup/backend.ai-webui/blob/main/react/src/components/ContainerRegistryList.tsx#L159

File: react/src/components/ImageInstallModal.tsx (hunk 24 above)
Comment (gahyuun) [reply react/src/components/ImageInstallModal.tsx]: I modified the rendering to only show the successful API image after clicking the install button. 
I made the requests in parallel and only updated the state of the successful images.  

https://github.com/lablup/backend.ai-webui/pull/2615/commits/ded9e4901c466f945092ca04d01f11ab7cd34d21

**Comment by rapsealk** [comment]
Please resolve the conflicts!

**Review by yomybaby — CHANGES_REQUESTED** [review]
I'm working to resolve blow list:
- [ ] Make the entire row clickable.
- [x] Resource Limitation Column: The rows in this table do not have a fixed height, so 'virtual' is not appropriate. Changing to a fixed height is not possible due to the resource limitation UI column.
- [ ] Base and Constraint Columns: Display them in one line.
- [ ] When you modify an image, the order can be changed. To prevent this, let's set the order using 'humanized_name'.

I'll share soon.

File: react/src/components/ImageInstallModal.tsx (hunk 22 above)
Comment (yomybaby) [reply react/src/components/ImageInstallModal.tsx]: @gahyuun  we can change this part using `upsertNotification` with `promise`. This is an [example](https://github.com/lablup/backend.ai-webui/blob/81c1dacb35a17b74da5c1e053097f511fd9e382d/react/src/pages/SessionLauncherPage.tsx#L502-L515).

**Review by yomybaby — CHANGES_REQUESTED** [review]
I got an error when I try to update using Manage app modal.  Please check it again.

**Review by yomybaby — APPROVED** [review]
LGTM

File: react/src/components/ImageList.tsx [hunk 28]
```diff
@@ -0,0 +1,512 @@
...
+              if (selectedRows.find((row) => row.id === record.id)) {
+                setSelectedRows((rows) =>
+                  rows.filter((row) => row.id !== record.id),
+                );
+              } else {
+                setSelectedRows((rows) => [...rows, record]);
+              }
+            },
+          })}
+          showSorterTooltip={false}
```
Comment (ironAiken2) [inline react/src/components/ImageList.tsx:475]:
```suggestion
          showSorterTooltip={false}
          sortDirections={['descend', 'ascend', 'descend']}
```
Logic to limit data sorting to ascending and descending order is not needed in the webui?

**Review by ironAiken2 — APPROVED** [review]
LGTM
